text/0000-00-00-coprocessor-write.md

```diff
@@ -0,0 +1,154 @@
+
+## Detailed design
+
+TiKV coprocessor has three types of endpoint requests now: DAG, ANALYZE and CHECKSUM. This RFC proposes to add two more types: DAG_PREWRITE and DAG_PESSIMISTIC_LOCK.
+
+### DAG_PREWRITE
+
+DAG_PREWRITE is a combination of a DAG request and a prewrite request. The whole transaction should span only one region.
+
+It first executes the DAG request to generate the rows to write. The rows are encoded to key-value pairs and prewritten. All the options of prewrite (except the start TS) are chosen by TiKV and will be returned in the coprocessor response.
+
+It is suitable for updating non-index fields of a table in an auto-commit transaction.
+
+#### Type ID
+
+DAG_PREWRITE sets the `tp` field to `110` in the coprocessor request. (106~109 are reserved for other read only requests.)
+
+#### Request format
+
+The `data` field in the request is a `DagRequest` message. But the `DagRequest` should follow these restrictions:
+
+* The `DagRequest` must be based on a `TableScan`.
+* The `DagRequest` must not contain aggregation executors.
+* The `DagRequest` must not change primary key columns.
+* The `DagRequest` should contain a `Projection` executor which retains the schema of the table.
+
+#### Response format
+
+The `data` field in the response contains a `PrewriteRequest` and a `PrewriteResponse`:
+
+```
+|len1|PrewriteReq|len2|PrewriteResp|
+|----|-----------|----|------------|
+
+len1:         length of PrewriteReq, 64-bit unsigned integer
+PrewriteReq:  encoded PrewriteRequest protobuf message
+len2:         length of PrewriteResp, 64-bit unsigned integer
+PrewriteResp: encoded PrewriteResponse protobuf message
+```
+
+`PrewriteReq` includes necessary information for the client to finish the 2PC:
+
+* List of keys
+* Primary key
+* Whether 1PC is used
+
+`PrewriteResp` is the response of the prewrite operation.
+
+Errors related to the execution of DAG request are set in the error fields in the coprocessor response. Prewrite related errors are set in `PrewriteResp`.
+
+#### Projection executor
+
+Now TiKV does not support the projection executor. We should have a basic implementation of it to support DAG_PREWRITE. 
+
+#### Future optimizations
+
+The initial implementation tends to be simple and general. So we reuse the current DAG execution and prewrite implementation. The problem is that we seek the key in the DAG execution and seek a second time in prewrite.
+
+If the update ranges are all point ranges, it is possible to lock these keys in the latch first and do the constraint checks of prewrite during the `TableScan`.
+
+This optimization needs some more work. We can create a `PointPrewrite` executor which is a specialized `TableScan` as the bottom executor of DAG. It yields results just like `TableScan` but also does constraint checks of prewrite. After the execution, we are safe to send the mutations to the underlying raft store without checking write conflicts again.
+
+### DAG_PESSIMISTIC_LOCK
```

Comment:
I think this feature doesn't have much benefit as DAG_PREWRITE.
It's only useful for UPDATE statements with primary key range condition.

Reply:
Could you clarify it a bit more? I don't see why it it's limited to the primary key range condition.
If it only has a selection exectuor, I think the filtered rows are exactly the rows that need to be locked.

Reply:
Sorry, you're right.

Reply:
But still, it's much more complicated than DAG_PREWRITE which only needs to handle the auto-commit single key non-index update.

It is unlikely to be implement in 5.0 release.

Reply:
Yes I agree it's more complicated than DAG_PREWRITE.

This document is only an RFC for technical discussion. I'm not going to decide which version will contain these features here.

I think it should be the product managers who decide it after the RFC is mostly accepted.

Reply:
How about we separate the two features into different RFCs?